### PR TITLE
feat(baselines) Add FEMNIST and reproduce results

### PR DESCRIPTION
**Pull request by Ashley0909**
### Description
The repository is limited to MNIST experiments, but there are multiple other datasets tested in the FedProx paper. A small issue in `docs/viz_and_plot_results.ipynb` is also fixed.

### Explanation
Loaded and Preprocess the FEMNIST dataset as mentioned in the paper -- subsampled 10 lower case characters ('a'-'j') and distribute only 5 classes to each device. Use `bash ./run_experiments.sh` to run and plot experimental results which match the results shown on the paper. The results are shown in the README.md file (at the bottom).

Also made a small change to the function viz() in `docs/viz_and_plot_results.ipynb`. Specifically, the first row in the picture are Losses and the second row are Accuracies, so it should be
`ax.set_ylabel("Loss" if row_idx == 0 else "Accuracy", fontsize=14)` instead of `col_idx == 0`.

### Checklist

- [x] Implement proposed change
- [ ] Write tests
- [x] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### Any other comments?
Some issues when undergoing CI checks but the issues are related to `flower/framework/py/flwr/common/record/arrayrecord.py`.

### CODE REVIEW

**Comment by jafermarq** [comment]
Hello @Ashley0909 , thanks for opening the PR to include the FEMNIST results. I skimmed through the changes and i note that your are setting 10 labels for FEMNIST, but it actually has 62...

**Comment by Ashley0909** [comment]
Hi @jafermarq ! Yes I was aware of this as well, but since the official FedProx paper subsampled the first 10 lower case characters instead of sampling all 62 classes (Section C.1) , I decided to replicate their implementation and reproduce their results.

**Comment by jafermarq** [comment]
> Hi @jafermarq ! Yes I was aware of this as well, but since the official FedProx paper subsampled the first 10 lower case characters instead of sampling all 62 classes (Section C.1) , I decided to replicate their implementation and reproduce their results.

You are right! I'll try to review your PR as soon as possible

**Comment by Ashley0909** [comment]
> > Hi @jafermarq ! Yes I was aware of this as well, but since the official FedProx paper subsampled the first 10 lower case characters instead of sampling all 62 classes (Section C.1) , I decided to replicate their implementation and reproduce their results.
> 
> You are right! I'll try to review your PR as soon as possible

Thank you!

**Comment by Ashley0909** [comment]
Hi @jafermarq , hope all is well! I was wondering if you've had a chance to review this PR? Any thoughts will be appreciated!

File: baselines/fedprox/README.md [hunk 1]
```diff
@@ -15,42 +14,57 @@ dataset: [MNIST]
...
-**What's implemented:** The code in this directory replicates the experiments in *Federated Optimization in Heterogeneous Networks* (Li et al., 2018) for MNIST, which proposed the FedProx algorithm. Concretely, it replicates the results for MNIST in Figure 1 and 7.
 
-**Datasets:** MNIST
+**What's implemented:** The code in this directory replicates the experiments in *Federated Optimization in Heterogeneous Networks* (Li et al., 2018) for MNIST and FEMNIST, which proposed the FedProx algorithm. Concretely, it replicates the results for both datasets in Figure 1 and 7.
+
+**Datasets:** MNIST, FEMNIST
 
 **Hardware Setup:** These experiments were run on a desktop machine with 24 CPU threads. Any machine with 4 CPU cores or more would be able to run it in a reasonable amount of time. Note: we install PyTorch with GPU support but by default, the entire experiment runs on CPU-only mode.
 
 **Contributors:** Charles Beauville, Javier Fernandez-Marques and Andrej Jovanović
```
Comment (jafermarq) [inline baselines/fedprox/README.md:25]: do you want to add your name at the end of this list?

File: baselines/fedprox/README.md [hunk 2]
```diff
@@ -6,7 +6,6 @@ dataset: [MNIST]
 ---
```
Comment (jafermarq) [inline baselines/fedprox/README.md:6]: Above this line we have some metadata indicating the datasets used, could you add `FEMNIST`?

File: baselines/fedprox/pyproject.toml
Comment (jafermarq) [inline baselines/fedprox/pyproject.toml:1]: I understand these changes have been done for the `FEMNIST` experiments. I think it would be better to keep it as is and then create a few config files for each of the hyperparam setups used to create the plot you include. One way to go ahead with this would be to create several config files as already done for `MNIST` inside the `conf/`. It might be better if you could move those to a subdirectory called `mnist` (note you'll need to adjust slighlty the `run_experiments.sh` script) and you create a new one for `femnist`. Finally, either add the femnist commands to the existing `run_experiments.sh` or create a new one exclusively for `FEMNIST`.

**Review by jafermarq — COMMENTED** [review]
Hi @Ashley0909 , thanks for the ping. Find below some comments. Nothing major just inviting your name to the list of _contributors_ in the `REAMDE.md` (if you want); and a suggestion to create config files so people can easily run the new `FEMNIST` experiments.

File: baselines/fedprox/pyproject.toml
Comment (jafermarq) [reply baselines/fedprox/pyproject.toml:1]: Open to other ways to go around this if it can be cleaner :)

File: baselines/fedprox/README.md (hunk 1 above)
Comment (Ashley0909) [reply baselines/fedprox/README.md]: Yes please, I have added my name at the end of the list, thank you!

**Comment by Ashley0909** [comment]
> Hi @Ashley0909 , thanks for the ping. Find below some comments. Nothing major just inviting your name to the list of _contributors_ in the `REAMDE.md` (if you want); and a suggestion to create config files so people can easily run the new `FEMNIST` experiments.

@jafermarq Thank you for inviting me as one of the contributors! I have also made some changes to the config files to make running the FEMNIST experiments easier.

File: baselines/fedprox/docs/viz_and_plot_femnist.ipynb
Comment (jafermarq) [inline baselines/fedprox/docs/viz_and_plot_femnist.ipynb:1]: This notebook is identical to the one for `MNIST` except the path used to load the results and the name given to the final image. Could we do the plotting in a single notebook? It's better to avoid duplicating code

File: baselines/fedprox/_static/FedProx_femnist.png
Comment (jafermarq) [inline baselines/fedprox/_static/FedProx_femnist.png:1]: I have re-run your experiments and mine looks quite different:
FedProx_femnist

Comment (jafermarq) [reply baselines/fedprox/_static/FedProx_femnist.png:1]: a bit more concerning is that `FedAvg` and `FedProx` with `mu=0.0` seem to converge very fast (a behaviour not seen in the Paper -- e.g. if you look at Fig7)

File: baselines/fedprox/conf/mnist/fedavg_sf_0.0.toml [hunk 3]
```diff
@@ -1,9 +1,15 @@
 [algorithm]
 name = "FedAvg"
+num-server-rounds = 100
+min-available-clients= 1000
+min-fit-clients= 1000
+num-clients = 1000
```
Comment (jafermarq) [inline baselines/fedprox/conf/mnist/fedavg_sf_0.0.toml:6]: why do we need to define these new entries in the MNIST configs? Note that `min-fit-clients` is not used anywhere

File: baselines/fedprox/conf/femnist/fedprox_mu_2.0_sf_0.5.toml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+[algorithm]
+name = "FedProx"
+num-server-rounds = 200
+min-available-clients= 200
+min-fit-clients= 200
+num-clients = 200
```
Comment (jafermarq) [inline baselines/fedprox/conf/femnist/fedprox_mu_2.0_sf_0.5.toml:6]: The number of clients in the simulation can't be changed on the fly with an argument. That's because this info only resides in the `options.num-supernodes = 1000` in the `pyproject.toml`. I'd recommend creating a new federation for `FEMNIST` given that it requires a different number of nodes/clients. See my comment in `pyproject.toml`

File: baselines/fedprox/pyproject.toml
Comment (jafermarq) [inline baselines/fedprox/pyproject.toml:1]: Create a new federation to be used in `FEMNIST` experiments like this (at the bottom of the file):
```
[tool.flwr.federations.femnist-federation]
options.num-supernodes = 200
options.backend.client-resources.num-cpus = 2
options.backend.client-resources.num-gpus = 0.0
```

Then, to use it, you'll need to adjust your `flwr run` commands to this:

```shell
flwr run . femnist-federation --run-config ....
```

File: baselines/fedprox/conf/femnist/fedprox_mu_2.0_sf_0.5.toml
Comment (jafermarq) [inline baselines/fedprox/conf/femnist/fedprox_mu_2.0_sf_0.5.toml:1]: The `Fedprox` configs need to update the `fraction-fit` so `20` clients out of the 200 are sampled. so set it to `fraction-fit = 0.1`

**Review by jafermarq — COMMENTED** [review]
Hi @Ashley0909 , thank you for incorporating the feedback. I re-ran your femnist script and got results quite different form the paper. But i think i found one point that could be the reason why: your experiments are still using the default 1000 nodes/clients as MNIST does. See my comment in the `pyproject.toml` and also those in the configs. In principle there shouldn't be a need to change the content of the `MNIST` configs.

File: baselines/fedprox/docs/viz_and_plot_femnist.ipynb
Comment (Ashley0909) [reply baselines/fedprox/docs/viz_and_plot_femnist.ipynb:1]: I have edited the ipynb so there is only one general viz_and_plot.ipynb which users change the dataset in the first cell under the variable DATASET

File: baselines/fedprox/conf/mnist/fedavg_sf_0.0.toml (hunk 3 above)
Comment (Ashley0909) [reply baselines/fedprox/conf/mnist/fedavg_sf_0.0.toml]: I have revert all the settings in pyproject.toml to be the original (MNIST) setting, and the .toml files under mnist/ folder are also revert back to original (other than the additional of path)

File: baselines/fedprox/conf/femnist/fedprox_mu_2.0_sf_0.5.toml (hunk 4 above)
Comment (Ashley0909) [reply baselines/fedprox/conf/femnist/fedprox_mu_2.0_sf_0.5.toml:6]: Thank you! I have added a new federation and changed the run command in the .sh file

File: baselines/fedprox/conf/femnist/fedprox_mu_2.0_sf_0.5.toml
Comment (Ashley0909) [reply baselines/fedprox/conf/femnist/fedprox_mu_2.0_sf_0.5.toml:1]: @jafermarq Thank you for pointing it out! I reread the FedProx and followed their experiment settings -- 10 clients selected (fraction-fit = 0.05). From the experiments you ran, I als realised that since the authors used learning rate = 0.003 for FEMNIST, unlike 0.03 for MNIST, therefore the convergence rate was unexpectedly fast. I have changed the configs in the /toml files under femnist/ folder and reran the experiments. The updated plot is saved as _static/FedProx_femnist.png and updated in README as well.

Comment (jafermarq) [reply baselines/fedprox/conf/femnist/fedprox_mu_2.0_sf_0.5.toml:1]: thanks for making the fixes! I'll re-review your changes soon!

Comment (Ashley0909) [reply baselines/fedprox/conf/femnist/fedprox_mu_2.0_sf_0.5.toml:1]: @jafermarq Thank you!

Comment (jafermarq) [reply baselines/fedprox/conf/femnist/fedprox_mu_2.0_sf_0.5.toml:1]: Hi @Ashley0909 , i'm sorry. I'm delayed getting back to you with some final feedback. I'll do so this week 💪

Comment (Ashley0909) [reply baselines/fedprox/conf/femnist/fedprox_mu_2.0_sf_0.5.toml:1]: Hi @jafermarq, no problem 💪🏻💪🏻 Hope to hear from you soon!
